Messaging System

njgrisafi opened the pull request:
Adds the private messaging system to the `wtf.core` and `wtf.api`

This is the basic implementation to do the following:
- Send a message
- Reply to a message
- Retrieve a message
- Retrieve a message replies
- Retrieve messages for an recipient

Part of #55

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,82 @@ def get_armor_by_id(armor_id):
+
+    $ curl \
+        --request GET \
+        --url http://localhost:5000/api/messages?recipient=<recipient_id>?status=<status> \
```
I think this should be (`&`):
```
...?recipient=<recipient_id>&status=<status>
```

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,82 @@ def get_armor_by_id(armor_id):
+        body=body.get('body'),
+        recipients=body.get('recipients')
+    ))
+    return jsonify({'message': message}), 200
```
I think this should be a `201 CREATED`

mrpudn commented on wtf/testing.py:
```diff
@@ -48,10 +48,12 @@ def post(self, path='', **kwargs):
     def get(self, path='', **kwargs):
         '''Send a GET request'''
         path = '%s%s' % (self.root_path, path)
+        query_string = kwargs.get('query_string', {})
```
Mind if we shorten this to like `query`? What do you think?

njgrisafi replied:
Yup I like that! I'll make the change

mrpudn commented on wtf/core/messages_test.py:
```diff
@@ -0,0 +1,153 @@
+def test_find_message_by_recipient_not_found():
+    expected = 'Recipient not found'
+    with pytest.raises(NotFoundError) as e:
+        messages.find_by_recipient(TEST_DATA['id'])
```
not sure if I'm reading this correctly: `TEST_DATA['id']` should be the message ID, right?

njgrisafi replied:
yup that's suppose to come from messages test data. Good catch!

njgrisafi replied:
Yup just made the fix to use recipients

mrpudn commented on wtf/core/messages_test.py:
```diff
@@ -0,0 +1,153 @@
+    mock_find_by_recipient.return_value = [
+        {'message': TEST_DATA['id'], 'status': 'unread'}
+    ]
+    actual = messages.get_recipient_messages()
```
How does this know which recipient to fetch methods for? Or is this just all messages?

njgrisafi replied:
I should've used the recipient ID. I updated this with my last commits

mrpudn commented on wtf/core/messages_test.py:
```diff
@@ -0,0 +1,153 @@
+    with pytest.raises(ValidationError) as e:
+        messages.validate({})
+    actual = e.value.errors
+    assert set(expected).issubset(actual)
```
Hey, regarding the validation stuff we talked about in gitter: don't worry about it for this PR. I'll probably come back and fix them all in a single PR later. :+1:

mrpudn commented on wtf/core/messages_test.py:
```diff
@@ -0,0 +1,153 @@
+    messages.validate({
+        'id': TEST_DATA['id'],
+        'parent': None,
+        'sender': None,
```
Sender should always have a value, right?

njgrisafi replied:
For now we don't associate anything with sender so that's why I left it as None. I can add in the test case just so we have it their for the future

mrpudn replied:
Oh, I thought it would have the character ID of the character sending the message?

njgrisafi replied:
@mrpudn ahh I thought that it was eventually going to be the account ID. Once we have the `auth` in place then we should be able to populate the `sender` value right?

mrpudn commented on wtf/core/messages_test.py:
```diff
@@ -0,0 +1,153 @@
+        'sender': None,
+        'subject': TEST_DATA['subject'],
+        'body': TEST_DATA['body'],
+        'copies': []
```
Copies shouldn't be empty, right?

njgrisafi replied:
I guess that makes sense. You can't send a message without a recipient. I'll add that check in the validation

mrpudn commented on wtf/core/messages_test.py:
```diff
@@ -0,0 +1,153 @@
+                'recipient': 'bar',
+                'status': 'unread'
+            }
+        ]
```
I think my explanation of message copies wasn't very clear. I meant for them to be like whole separate entities, not embedded objects.

Imagine how you'd represent them in a relational database fashion. `messages` table, related to `message_copies` table by `messages.id` equal to `message_copies.message`.

mrpudn replied:
well actually, they do **become** embedded objects, but only after the `transform()` part.. actually still not totally clear on that myself

njgrisafi replied:
Ahh okay yeah I didn't think embedding the same data twice made much sense.

Do you want me to make a new file `core/message_copies.py` or keep message copies in the `core/messages.py`?

I think you're saying you just want me to add another repo in `core/messages.py` for `message_copies_by_message_id` . Then when a user requests `/message/<message_id>`, we fetch the message copies, then transform the data to include the copies and return the transformed data to the user.

Or do we only want to return the array of recipient addresses in `message.copies`?

Sorry if this is confusing, I'm bit lost right now haha

mrpudn replied:
haha no worries - we can probably remove the message copies stuff and tackle that in a second PR if that's not too difficult. I'll put some thoughts into the issue for this.

njgrisafi replied:
I think we can get it done in this PR. Your last update on the issue help clear things up

njgrisafi commented:
@mrpudn Updated this PR with the following:

- Refactored message replies into separate repo
- Added transform method for message and message_copies
- Added retrieving replies for a message

I'm thinking we should create create a `save_copies` method because the message `save` method is getting large and maybe out of it's scope.

Let me know if you catch anything else I should change. I think some of the naming and error messages should be refactored at some point too.

mrpudn commented:
Yup, go ahead with `save_copy()` or `save_copies()` (or both, even) - there's a `save_recipe()` and a `save()` method in `weapons` and `armor`, and I'm about to add a `save_token()` to the `accounts` module, so totally fits the pattern.

mrpudn commented on wtf/api/routes.py:
```diff
@@ -25,6 +25,11 @@ def get_json_body():
 
+def get_query_args():
+    '''Get the request query args.'''
+    return request.args
```
Any reason there's a method for this? :P

Seem like we could just use `request.args` on its own - were you thinking we might add something to this in the future?

njgrisafi replied:
Yeah no good reason, I'll go ahead and remove this. I was thinking to keep it consistent since many other function may be accessing the same data and we were doing something similar for the body.

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
+
+    $ curl \
+        --request GET \
+        --url http://localhost:5000/api/messages?recipient=<recipient_id>&status=<status> \
```
We'll have to figure out a better way to represent this in the future (probably via `--data-urlencode`), since there could be a lot of potential query args one could use. (No need to tackle that in this PR, though)

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
+    args = get_query_args()
+    status = args.get('status')
+    recipient = args.get('recipient')
+    message_copies = messages.get_recipient_message_copies(recipient=recipient, status=status)
```
How about maybe `find_by_recipient_id()` to match the convention elsewhere (`accounts.find_by_email()`, `characters.find_by_id()`, etc.)?

njgrisafi replied:
@mrpudn I already added a method for `find_copies_by_recipient_id()`

This method `get_recipient_message_copies()`  calls `find_copies_by_recipient_id()` and filters the copies by the status (if status is set).

Should I just add the filtering logic in the `find_copies_by_recipient_id()` method?

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
+    status = args.get('status')
+    recipient = args.get('recipient')
+    message_copies = messages.get_recipient_message_copies(recipient=recipient, status=status)
+    return jsonify({'messages': messages.transform_copies(message_copies)}), 200
```
I don't think we've dealt with multiple things getting transformed - how about `transform_all()` since this should transform multiple raw messages and add their message copies.

What gets passed in here should be an array of messages. The `transform_all()` will fetch the message copies for you.

njgrisafi replied:
@mrpudn I think that's a great idea! I'll add that method and simplify this logic

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
+        body=body.get('body'),
+        recipients=body.get('recipients')
+    ))
+    return jsonify({'message': message}), 201
```
This should be a `transform()`-ed message

njgrisafi replied:
Yeah I wasn't sure what to put here. Since you created the message I thought just to send back the raw inserted object. But yeah I'll switch it back to return the transformed value.

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
+
+
+@BLUEPRINT.route('/messages/<message_id>', methods=['GET'])
+def get_message_by_id(message_id):
```
`find_by_id` (since the module is already called `messages`, no need to add it here)

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
+        --data '{
+            "subject": "...",
+            "body": "...",
+            "recipients": ["..."],
```
no comma here :)

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
+        recipients=body.get('recipients'),
+        parent=message_id
+    ))
+    return jsonify({'message': messages.transform(message)}), 201
```
You know, for this route, I think we can reuse the `create_message()` handler and make `message_id` optional (i.e. `create_message(message_id=None)`.

See: https://stackoverflow.com/questions/14023664/flask-url-route-route-several-urls-to-the-same-function

Pretty cool, right? :)

mrpudn commented on wtf/api/routes.py:
```diff
@@ -324,3 +329,98 @@ def get_armor_by_id(armor_id):
+
+
+@BLUEPRINT.route('/messages/<message_id>/replies', methods=['GET'])
+def get_message_replies(message_id):
```
What's this for again? Wouldn't you just use `GET /api/messages`? I haven't thought through fetching an entire reply chain yet

njgrisafi replied:
I thought it would be useful to retrieve all replies for a given parent message but I guess you can traverse the messages endpoint using parent value.

I'll remove this function.

mrpudn commented on wtf/core/messages.py:
```diff
@@ -0,0 +1,208 @@
+Messages have the following properties:
+  * id: the message UUID (Universally Unique Identifier)
+  * body: the body of the message
+  * sender: the account id that send the messages
```
I think this will be the character ID, not the account ID

njgrisafi replied:
Okay I refactor that to be character ID

mrpudn commented on wtf/core/messages.py:
```diff
@@ -0,0 +1,208 @@
+  * sender: the account id that send the messages
+  * parent: the previous message id (None if no previous message)
+  * subject: the subject of the message
+  * copies: the recipients copies of the message
```
This should be broken out into another entity with its own properties

mrpudn commented on wtf/core/messages.py:
```diff
@@ -0,0 +1,208 @@
+        parent: the UUID for the previous message
+        subject: the subject of the message
+        body: the body of the message
+        recipients: the list of UUID for the recipients
```
You don't need this (line 25 to line 31) since it's already in the module doc comment. Duplication :)

mrpudn commented on wtf/core/messages.py:
```diff
@@ -0,0 +1,208 @@
+    body = kwargs.get('body')
+    subject = kwargs.get('subject', '(No Subject)')
+    recipients = kwargs.get('recipients')
+    recipients = recipients if recipients is not None else []
```
you can replace these two lines with: 
```py
recipients = kwargs.get('recipients', [])
```

njgrisafi replied:
@mrpudn this wasn't working because the key exists and the value is set to None. This is because the calling function sets the default parameter to None. Since the key exist it will return the value (which could be none).

I guess we can avoid this by not setting the parameter if it is None.

https://www.tutorialspoint.com/python/dictionary_get.htm

njgrisafi replied:
We can do `recipients = kwargs.get('recipients', [])` when we process the request and then call this function.

mrpudn replied:
oh gotcha, then you should be able to do this:
```py
recipients = kwargs.get('recipients') or []
```
Short and sweet :)

mrpudn commented on wtf/core/messages.py:
```diff
@@ -0,0 +1,208 @@
+        'created_at': datetime.utcnow().isoformat()
+    }
+    for recipient in recipients:
+        message['copies'].append(create_copy(message, recipient=recipient))
```
I think message copies is out of scope for this function and should not be nested like this. Let's continue discussion on this one in the issue thread (#92).

njgrisafi replied:
Yup that's what I mentioned above. We can definitely break out this and the save logic for the message copies.

mrpudn commented on wtf/core/messages.py:
```diff
@@ -0,0 +1,208 @@
+def create_copy(message, **kwargs):
+    '''Create a message copy
+
+    MessageCopies have the following properties:
```
Not a big deal, but I think we should go with "Message Copies" instead of like a Java class "MessageCopies" to describe this.

mrpudn commented on wtf/core/messages.py:
```diff
@@ -0,0 +1,208 @@
+        recipient: the UUID of the recipient
+        status: the status of the current message: unread, read, deleted, saved
+        read_at: the time the recipient read the mesaage
+        deleted_at: the time the recipient deleted the message
```
Also, let's put this stuff in the module comment instead of here :)

mrpudn commented on wtf/core/messages.py:
```diff
@@ -0,0 +1,208 @@
+def save(message):
+    '''Persist a message.
+
+    Creates a message
```
No need for this line, you can just do `'''Persist a message.'''`

njgrisafi commented:
@mrpudn updated this branch. Let me know if you catch anything else you want me to change.

mrpudn commented:
Hey, I'm going to go ahead and merge this. I'll probably create a follow up PR with a few tweaks.
